File: composey/models/compose.py
from typing import Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .semantic import Capability, Ingress


class Port(BaseModel):
    mode: Literal["ingress", "egress"] = "ingress"
    target: int
    published: Optional[int] = None
    protocol: Literal["tcp", "udp"] = "tcp"


class Build(BaseModel):
    """
    docker-compose build

    `context` is expected to be relative to the compose file. The parser makes
    it so; taking the basename of an absolute path (as this model used to) is
    wrong whenever the context is not directly beside the compose file, which
    is the normal case in a monorepo where several services build from the root.
    """

    context: str = Field(description="context")
    dockerfile: Optional[str] = Field(
        default=None, description="Dockerfile path, relative to the context"
    )


class Dependency(BaseModel):
    condition: str = Field(description="condition")
    required: bool = Field(description="required", default=True)


class SecretDefinition(BaseModel):
    source: str
    target: Optional[str] = None
    uid: Optional[str] = None
    gid: Optional[str] = None
    mode: Optional[int] = None


class VolumeDefinition(BaseModel):
    type: str
    source: Optional[str] = None
    target: str
    read_only: bool = False


class HealthCheckConfig(BaseModel):
    """Health check configuration for x-composey."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["http", "tcp"] = Field(default="http")
    path: str = Field(default="/")
    port: Optional[int] = Field(default=None)


class AutoScalingMetricConfig(BaseModel):
    """Auto-scaling metric configuration for x-composey."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["cpu", "memory", "requests_per_target"] = Field(
        default="cpu",
        description="Metric to scale on",
    )
    target: float = Field(
        default=70.0,
        gt=0,
        le=100,
        description="Target value (percentage for cpu/memory, count for requests)",
    )


class AutoScalingConfig(BaseModel):
    """Auto-scaling configuration for x-composey."""

    model_config = ConfigDict(extra="forbid")

    metrics: list[AutoScalingMetricConfig] = Field(
        default_factory=lambda: [
            AutoScalingMetricConfig(type="cpu", target=70.0),
            AutoScalingMetricConfig(type="memory", target=80.0),
        ],
    )
    scale_in_cooldown: int = Field(default=300, ge=0)
    scale_out_cooldown: int = Field(default=60, ge=0)


class XComposey(BaseModel):
    """
    The `x-composey` block on a service.

    Validated with `extra="forbid"` on purpose. An override you can misspell is
    not an override: before this existed, `capabilty: database` was silently
    dropped and the service deployed as whatever the compiler guessed.
    """

    model_config = ConfigDict(extra="forbid")

    capability: Optional[Capability] = Field(
        default=None,
        description="What this service really is, when the image name does not say",
    )
    ingress: Optional[Ingress] = Field(
        default=None,
        description="How this service is reached from outside: path, port, health",
    )
    health_check: Optional[HealthCheckConfig] = Field(
        default=None,
        description="Health check configuration. Defaults to HTTP / on the ingress port.",
    )
    auto_scaling: Optional[AutoScalingConfig] = Field(
        default=None,
        description="Auto-scaling configuration. When min_scale != max_scale, "
        "auto-scaling is enabled. Defaults to CPU 70% and Memory 80%.",
    )

    @model_validator(mode="before")
    @classmethod
    def _bare_ingress_means_defaults(cls, data):
        """
        `ingress:` with nothing under it declares a default route.

        Without this it parses as null and the service is quietly internal —
        reintroducing, at the only place it still could, exactly the silent
        non-exposure this design exists to prevent.
        """
        if isinstance(data, dict) and "ingress" in data and data["ingress"] is None:
            data = {**data, "ingress": {}}
        return data

    @property
    def exposure(self) -> Optional[Ingress]:
        """The route this service declares, if any."""
        return self.ingress

    size: Literal["small", "medium", "large"] = Field(default="small")
    cpu: Optional[int] = Field(default=None, gt=0)
    memory: Optional[int] = Field(default=None, gt=0)
    min_scale: int = Field(default=1, ge=0)
    max_scale: int = Field(default=1, ge=1)
    schedule: Optional[str] = Field(default=None)
    cdn: bool = Field(default=False)
    startup_grace_period: Optional[int] = Field(default=None, ge=0)
    health_check_grace_period: Optional[int] = Field(
        default=None, ge=0, description="Deprecated spelling of startup_grace_period"
    )

    @property
    def grace_period(self) -> Optional[int]:
        """The startup grace period under either spelling."""
        if self.startup_grace_period is not None:
            return self.startup_grace_period
        return self.health_check_grace_period


class Service(BaseModel):
    """
    docker-compose service
    """

    model_config = ConfigDict(populate_by_name=True, extra="allow")

    build: Optional[Build] = Field(description="build", default=None)
    ports: Optional[list[Port]] = Field(description="ports", default=None)
    image: Optional[str] = Field(description="image", default=None)
    command: Optional[Union[str, list[str]]] = Field(
        description="command", default=None
    )
    environment: dict[str, Optional[str]] = Field(
        description="environment", default_factory=dict
    )
    depends_on: dict[str, Dependency] = Field(default_factory=dict)
    platform_env: list[str] = Field(
        default_factory=list,
        description="Variables the service needs whose values come from the "
        "platform rather than the compose file. Populated by the parser.",
    )
    networks: Optional[dict[str, Optional[dict]]] = Field(
        default=None,
        description="Networks this service joins. `docker compose config` always "
        "resolves this, materialising `default` when the file declares none.",
    )
    secrets: Optional[list[Union[str, SecretDefinition]]] = Field(default=None)
    volumes: Optional[list[Union[str, VolumeDefinition]]] = Field(default=None)

    @property
    def x_composey_raw(self) -> dict:
        return (self.model_extra or {}).get("x-composey") or {}


class NetworkDefinition(BaseModel):
    model_config = ConfigDict(extra="allow")

    name: Optional[str] = None
    external: bool = False


class Application(BaseModel):
    model_config = {"extra": "ignore"}

    services: dict[str, Service]
    networks: dict[str, Optional[NetworkDefinition]] = Field(default_factory=dict)

File: composey-go/internal/models/compose.go
package models

import (
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
)

type ComposeApplication struct {
	Services map[string]ComposeService     `json:"services,omitempty"`
	Networks map[string]*NetworkDefinition `json:"networks,omitempty"`
	Volumes  map[string]interface{}        `json:"volumes,omitempty"`
	Secrets  map[string]ComposeSecret      `json:"secrets,omitempty"`
}

type ComposeService struct {
	Image       string                 `json:"image,omitempty"`
	Build       *BuildConfig           `json:"build,omitempty"`
	Ports       []PortConfig           `json:"ports,omitempty"`
	Environment map[string]*string     `json:"environment,omitempty"`
	DependsOn   map[string]Dependency  `json:"depends_on,omitempty"`
	Networks    map[string]interface{} `json:"networks,omitempty"`
	Volumes     []VolumeDefinition     `json:"volumes,omitempty"`
	Secrets     []interface{}          `json:"secrets,omitempty"`
	Command     interface{}            `json:"command,omitempty"`
	XComposey   interface{}            `json:"x-composey,omitempty"`
	PlatformEnv []string               `json:"platform_env,omitempty"`
}

func (s *ComposeService) GetNetworks() []string {
	if len(s.Networks) == 0 {
		return nil
	}
	networks := make([]string, 0, len(s.Networks))
	for name := range s.Networks {
		networks = append(networks, name)
	}
	return networks
}

type BuildConfig struct {
	Context    string   `json:"context,omitempty"`
	Dockerfile string   `json:"dockerfile,omitempty"`
	Args       []string `json:"args,omitempty"`
	Target     string   `json:"target,omitempty"`
}

type PortConfig struct {
	Target    uint32 `json:"target"`
	Published string `json:"published,omitempty"`
	Protocol  string `json:"protocol,omitempty"`
}

type Dependency struct {
	Condition string `json:"condition,omitempty"`
	Required  bool   `json:"required,omitempty"`
}

type ComposeSecret struct {
	File string `json:"file,omitempty"`
}

type NetworkDefinition struct {
	Name     string `json:"name,omitempty"`
	External bool   `json:"external,omitempty"`
}

type VolumeDefinition struct {
	Type     string `json:"type"`
	Source   string `json:"source,omitempty"`
	Target   string `json:"target"`
	ReadOnly bool   `json:"read_only,omitempty"`
}

// XComposey is the `x-composey` block on a service.
//
// Ported from models/compose.py's XComposey, which validates with Pydantic's
// extra="forbid" plus per-field bounds (Field(ge=..., gt=...)). Go's
// encoding/json has no equivalent declarative validation, so both are
// enforced by hand in UnmarshalJSON — deliberately, since the failure this
// exists to prevent is exactly a misspelled key or an out-of-range value
// being silently accepted rather than rejected: `capabilty: database` was
// once silently dropped, and the service deployed as whatever the compiler
// guessed from the image name instead.
type XComposey struct {
	Capability             *Capability               `json:"capability,omitempty"`
	Ingress                *IngressConfig            `json:"ingress,omitempty"`
	HealthCheck            *HealthCheckConfig        `json:"health_check,omitempty"`
	AutoScaling            *ComposeAutoScalingConfig `json:"auto_scaling,omitempty"`
	Size                   string                    `json:"size,omitempty"`
	CPU                    *int                      `json:"cpu,omitempty"`
	Memory                 *int                      `json:"memory,omitempty"`
	MinScale               int                       `json:"min_scale"`
	MaxScale               int                       `json:"max_scale"`
	Schedule               string                    `json:"schedule,omitempty"`
	CDN                    bool                      `json:"cdn,omitempty"`
	StartupGracePeriod     *int                      `json:"startup_grace_period,omitempty"`
	HealthCheckGracePeriod *int                      `json:"health_check_grace_period,omitempty"`
}

var validCapabilities = map[string]bool{
	"container": true, "database": true, "cache": true, "object-storage": true,
}

var validSizes = map[string]bool{"small": true, "medium": true, "large": true}

// intOrString accepts a JSON number or a numeric JSON string ("5"), matching
// Pydantic's lenient coercion — docker-compose users occasionally quote
// numeric x-composey values, and Python accepted that silently. Go's
// encoding/json does not coerce string-to-int by default, and returning that
// error unchanged (rather than deliberately re-implementing the coercion
// Pydantic already did) turns any quoted number into a hard, surprising
// break for a file that used to compile.
func intOrString(raw json.RawMessage, field string) (int, error) {
	var asInt int
	if err := json.Unmarshal(raw, &asInt); err == nil {
		return asInt, nil
	}
	var asString string
	if err := json.Unmarshal(raw, &asString); err == nil {
		if n, err := strconv.Atoi(asString); err == nil {
			return n, nil
		}
	}
	return 0, fmt.Errorf("%s must be an integer, got %s", field, string(raw))
}

// UnmarshalJSON enforces what Pydantic's extra="forbid" and Field bounds did
// on the Python model: unknown keys and out-of-range values are rejected
// outright rather than silently ignored or clamped.
func (x *XComposey) UnmarshalJSON(data []byte) error {
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}

	known := map[string]bool{
		"capability": true, "ingress": true, "health_check": true,
		"auto_scaling": true, "size": true, "cpu": true, "memory": true,
		"min_scale": true, "max_scale": true, "schedule": true, "cdn": true,
		"startup_grace_period": true, "health_check_grace_period": true,
	}
	unknown := make([]string, 0)
	for key := range raw {
		if !known[key] {
			unknown = append(unknown, key)
		}
	}
	if len(unknown) > 0 {
		sort.Strings(unknown)
		return fmt.Errorf("unknown field(s): %v", unknown)
	}

	result := XComposey{Size: "small", MinScale: 1, MaxScale: 1}

	if v, ok := raw["capability"]; ok {
		var s string
		if err := json.Unmarshal(v, &s); err != nil {
			return err
		}
		if !validCapabilities[s] {
			return fmt.Errorf("capability must be one of container, database, cache, object-storage, got %q", s)
		}
		c := Capability(s)
		result.Capability = &c
	}
	if v, ok := raw["ingress"]; ok {
		if string(v) == "null" {
			// Bare `ingress:` with nothing under it declares a default
			// route. Left as null it parses as no ingress at all, quietly
			// making the service internal — reintroducing, at the only
			// place it still could, exactly the silent non-exposure this
			// design exists to prevent.
			result.Ingress = &IngressConfig{}
		} else {
			result.Ingress = &IngressConfig{}
			if err := json.Unmarshal(v, result.Ingress); err != nil {
				return err
			}
		}
	}
	if v, ok := raw["health_check"]; ok {
		result.HealthCheck = &HealthCheckConfig{}
		if err := json.Unmarshal(v, result.HealthCheck); err != nil {
			return err
		}
	}
	if v, ok := raw["auto_scaling"]; ok {
		result.AutoScaling = &ComposeAutoScalingConfig{}
		if err := json.Unmarshal(v, result.AutoScaling); err != nil {
			return err
		}
	}
	if v, ok := raw["size"]; ok {
		if err := json.Unmarshal(v, &result.Size); err != nil {
			return err
		}
		if !validSizes[result.Size] {
			return fmt.Errorf("size must be one of small, medium, large, got %q", result.Size)
		}
	}
	if v, ok := raw["cpu"]; ok {
		n, err := intOrString(v, "cpu")
		if err != nil {
			return err
		}
		if n <= 0 {
			return fmt.Errorf("cpu must be greater than 0, got %d", n)
		}
		result.CPU = &n
	}
	if v, ok := raw["memory"]; ok {
		n, err := intOrString(v, "memory")
		if err != nil {
			return err
		}
		if n <= 0 {
			return fmt.Errorf("memory must be greater than 0, got %d", n)
		}
		result.Memory = &n
	}
	if v, ok := raw["min_scale"]; ok {
		n, err := intOrString(v, "min_scale")
		if err != nil {
			return err
		}
		if n < 0 {
			return fmt.Errorf("min_scale must be greater than or equal to 0, got %d", n)
		}
		result.MinScale = n
	}
	if v, ok := raw["max_scale"]; ok {
		n, err := intOrString(v, "max_scale")
		if err != nil {
			return err
		}
		if n < 1 {
			return fmt.Errorf("max_scale must be greater than or equal to 1, got %d", n)
		}
		result.MaxScale = n
	}
	if v, ok := raw["schedule"]; ok {
		if err := json.Unmarshal(v, &result.Schedule); err != nil {
			return err
		}
	}
	if v, ok := raw["cdn"]; ok {
		if err := json.Unmarshal(v, &result.CDN); err != nil {
			return err
		}
	}
	if v, ok := raw["startup_grace_period"]; ok {
		n, err := intOrString(v, "startup_grace_period")
		if err != nil {
			return err
		}
		if n < 0 {
			return fmt.Errorf("startup_grace_period must be greater than or equal to 0, got %d", n)
		}
		result.StartupGracePeriod = &n
	}
	if v, ok := raw["health_check_grace_period"]; ok {
		n, err := intOrString(v, "health_check_grace_period")
		if err != nil {
			return err
		}
		if n < 0 {
			return fmt.Errorf("health_check_grace_period must be greater than or equal to 0, got %d", n)
		}
		result.HealthCheckGracePeriod = &n
	}

	*x = result
	return nil
}

func (x *XComposey) GetGracePeriod() *int {
	if x.StartupGracePeriod != nil {
		return x.StartupGracePeriod
	}
	return x.HealthCheckGracePeriod
}

type IngressConfig struct {
	Path        string            `json:"path"`
	Port        *int              `json:"port,omitempty"`
	HealthCheck HealthCheckConfig `json:"health_check,omitempty"`
}

type HealthCheckConfig struct {
	Type string `json:"type,omitempty"`
	Path string `json:"path,omitempty"`
	Port *int   `json:"port,omitempty"`
}

type ComposeAutoScalingConfig struct {
	Metrics          []AutoScalingMetricConfig `json:"metrics,omitempty"`
	ScaleInCooldown  int                       `json:"scale_in_cooldown,omitempty"`
	ScaleOutCooldown int                       `json:"scale_out_cooldown,omitempty"`
}

type AutoScalingMetricConfig struct {
	Type   string  `json:"type,omitempty"`
	Target float64 `json:"target,omitempty"`
}
